### PR TITLE
Proposal: New logging driver that logs to tcp, udp and unix domain sockets

### PR DESCRIPTION
**Pull request by adamtulinius**
I'm proposing to add a new logging driver, that can log to a tcp, udp or unix domain socket.

The purpose is to make it easy to do custom processing of logs: Instead of writing yet another logging driver for Docker, this feature would make it easy to write a separate daemon to process logs.

A range of formats could be supported, but I suggest starting with a json-based format, where each logentry is delimited with \n, inspired by the [Logstash json_lines codec](https://www.elastic.co/guide/en/logstash/current/plugins-codecs-json_lines.html) and the [Logstash json format](https://gist.github.com/jordansissel/2996677). 
The resulting log entires will thus end up looking like:

```
{"@timestamp":"2015-11-16T01:23:00.233368446Z","@version":1,"hostname":"foo.example.com","containerID":"fe6bf90dcde1420588fded941fc4ed0122640723d893af53aa2449deeb4d1a73","message":"Hello from Docker.","source":"stdout"}
```

The reason for suggesting this particular format, is that I assume more people are familiar with JSON than many of the alternatives (such as MessagePack), and that this format is already used by Logstash.

I would like to give implementing it a shot on my own. I already made a proof-of-concept implementation of the current proposal.

ping @cpuguy83

### CODE REVIEW

**Comment by LK4D4** [comment]
However I think it's good, I'm not sure that we should use format with `@` for this. But it's just my opinion. I think we can start work on code. WDYT @cpuguy83 ?

**Comment by cpuguy83** [comment]
Also thinking msgpack or protobuf.

**Comment by dqminh** [comment]
I'm not too fond of yet another (JSON) log format. We already have many log drivers that outputs processable logs, why do we want to have another one ?

**Comment by cpuguy83** [comment]
@dqminh it's basically to make this "pluggable" so we don't have to keep adding more and more drivers to the codebase.

**Comment by dqminh** [comment]
@cpuguy83 i would argue that any of the current structured-based log driver ( gelf / fluentd / even journald ) is already pluggable, although each having pros/cons of their own.

If we decide to have another pluggable one, i would like to see its benefits outweighs the rests.

**Comment by cpuguy83** [comment]
@dqminh I think there is value in being able to blast out the logs in the "native" structure we have, (logging) service providers can implement their own adapters without being bound by shortcomings in the currently supported formats.

Most logging services implements one (or likely multiple) formats (like syslog and journald) so that they can consume logs from just about any system that exists, but these are lowest common denominator, which is really icky.

**Comment by adamtulinius** [comment]
> However I think it's good, I'm not sure that we should use format with @ for this. But it's just my opinion. I think we can start work on code. WDYT @cpuguy83 ?

Actually it would probably be better to *not* emulate how logstash does it, especially if we're also going to support other output formats, such as msgpack/protobufs. That way we could try to make the structure of the data similar.

> Also thinking msgpack or protobuf.

Would be pretty trivial to support different formats. Personally I think it's worth supporting JSON because the barrier to entry is really low.

> @dqminh I think there is value in being able to blast out the logs in the "native" structure we have, (logging) service providers can implement their own adapters without being bound by shortcomings in the currently supported formats.

Exactly. I feel something like this should just dump as much data as possible, with the least amount of processing, and basically not be very opinionated about how the data should be used. 

Thanks for the input so far.

**Comment by thaJeztah** [comment]
Looks like we're okay implementing this, @cpuguy83 @LK4D4?

@adamtulinius are you already working on the code for this?

**Comment by adamtulinius** [comment]
> @adamtulinius are you already working on the code for this?
Yes, I have quite a lot of it implemented, I just chose not to push it before reaching some sort of consensus about what to do.

I should be able to get it rebased on current master and add it to the PR during this weekend.

**Comment by thaJeztah** [comment]
Thanks @adamtulinius, we can continue discussing the details when the code is there

**Comment by thaJeztah** [comment]
Also note this proposal btw; https://github.com/docker/docker/issues/18604

**Comment by thaJeztah** [comment]
ping @adamtulinius, have you had time to look at this?

**Comment by adamtulinius** [comment]
Sorry about the delay. Had managed to break my dev environment.
Anyways, i've pushed an implementation that is basically working, but I'm going to clean up the code the following days.
Also, I'm in the process of implementing tests.

**Comment by adamtulinius** [comment]
I thought about something, though:
When log messages can't be delievered, what should the driver do?
1) drop messages and print an error
2) attempt delivery later on, either storing messages for X minutes, or storing up to Y messages for later delivery

Thoughts?

**Comment by thaJeztah** [comment]
Thanks @adamtulinius !

**Comment by thaJeztah** [comment]
ping @LK4D4 @cpuguy83 can you have a look at this, perhaps some thoughts on https://github.com/docker/docker/pull/18001#issuecomment-166171647 ?

**Comment by thaJeztah** [comment]
ping @cpuguy83 @LK4D4 can you have a look please?

**Comment by cpuguy83** [comment]
Looks good.

I would also want to make sure we can customize the fields that get sent much like we do with a couple of the other drivers. If not right away, at least in a way that we don't have to make a breaking change to the outgoing message format.

I still think we need to look at either using msgpack or protobuf to encode the messages... and of course json can be benched as well.
We need to make sure we can get these messages out of docker as quickly and w/ as low impact to the host as possible.

**Comment by tiborvass** [comment]
overall +1 on the design, but am wondering what are the drawbacks? how would performance be impacted? /cc @adamtulinius

**Comment by icecrime** [comment]
Ping @adamtulinius @LK4D4: shall we move to code review?

**Comment by cpuguy83** [comment]
I've created some (naive) benchmarks for encoding log messages for reference:
```
BenchmarkJSON-8    	 1000000	      2254 ns/op	     568 B/op	       8 allocs/op
BenchmarkProtobuf-8	 2000000	       602 ns/op	     336 B/op	       5 allocs/op
BenchmarkMsgpack-8 	 1000000	      2037 ns/op	     384 B/op	       6 allocs/op
BenchmarkFlatbuf-8 	 5000000	       303 ns/op	       0 B/op	       0 allocs/op
```

https://gist.github.com/cpuguy83/ac5d8f87bce9c37c0009

**Comment by LK4D4** [comment]
@cpuguy83 hell, flatbuffers is pretty cool :)

**Comment by tiborvass** [comment]
<3 flatbuffers

**Comment by LK4D4** [comment]
Let's move to code-review, it seems pretty easy to extend later.

File: daemon/logger/socket/socket.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
+// +build linux
```
Comment (LK4D4) [inline daemon/logger/socket/socket.go]: Don't see why it should work only on linux.

File: daemon/logger/socket/socket.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+func (s *socketLogger) Log(msg *logger.Message) error {
+	res := message{
+		Timestamp:   msg.Timestamp,
+		Hostname:    "",
+		ContainerID: msg.ContainerID,
+		Message:     string(msg.Line),
+		Source:      msg.Source,
+	}
+
+	json, err := json.Marshal(res)
```
Comment (LK4D4) [inline daemon/logger/socket/socket.go]: Good optimization would be to create persistent `json.Encoder` with conn in the logger object. Also `json.Encoder` automatically writes `\n`.

File: daemon/logger/socket/socket.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+	return nil
+}
+
+func (s *socketLogger) Name() string {
+	return name
+}
+
+func (s *socketLogger) Close() error {
+	if s.connection != nil {
```
Comment (LK4D4) [inline daemon/logger/socket/socket.go]: You probably forgot Lock here.

File: daemon/logger/socket/socket.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		err := s.connection.Close()
+		if err != nil {
```
Comment (LK4D4) [inline daemon/logger/socket/socket.go]: can be just `return s.connection.Close()`

**Comment by cpuguy83** [comment]
And was able to get protbuf a bit better than flatbuff with some tweaks to use an encoder instead of calling `Marshal` directly:
```
BenchmarkProtobuf-8	 5000000	       244 ns/op	       0 B/op	       0 allocs/op
BenchmarkFlatbuf-8 	 5000000	       312 ns/op	       0 B/op	       0 allocs/op
```

**Comment by aaronlehmann** [comment]
@cpuguy83 Check out https://github.com/alecthomas/go_serialization_benchmarks/blob/master/README.md

**Comment by thaJeztah** [comment]
ping @adamtulinius looks like this needs a rebase; also my can you address the comments so far?

**Comment by thaJeztah** [comment]
ping @adamtulinius can you rebase and address the comments?

**Comment by thaJeztah** [comment]
ping @cpuguy83 @LK4D4 should we carry this?

**Comment by thaJeztah** [comment]
Maintainers meeting today;

Although we like this feature, we should look at this together with the general "logging plugin" proposal https://github.com/docker/docker/issues/18604. There's really an overlap in the two, and we should decide what kind of mechanism we want to use to have external plugins. We should be careful developing our own protocols, unless there are no other options.

It's possible that we arrive at the solution in this PR, but moving this forward without having a decision on the general logging plugin system would be bad, as we may end up with two things doing exactly the same.

In addition, we have the GELF driver, which *is* a standard solution and already can be used for many purposes that this PR tries to resolve https://docs.docker.com/engine/admin/logging/overview/#gelf-options

For now, I'm going to close this PR until we have agreement on the plugin design. Thanks so much for starting the work on this.

**Comment by subfuzion** [comment]
I've proposed a solution that lets docker stream the logs to an image you specific with "custom" logging driver [here](https://github.com/docker/docker/issues/18604#issuecomment-198786889).

**Pull request closed without merging**